File: Components/DevExpressVCL/Demos/ExpressEditors Library/Delphi/ActivityIndicatorDemo/ActivityIndicatorDemoMain.pas
unit ActivityIndicatorDemoMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Forms, ComCtrls, BaseForm, Menus, Classes, Controls, StdCtrls, cxGraphics, cxControls,
  cxLookAndFeels, cxLookAndFeelPainters, cxPC, cxContainer, cxEdit, cxGroupBox, dxActivityIndicator,
  cxLabel, cxTextEdit, cxMaskEdit, cxSpinEdit, dxBevel, cxDropDownEdit, cxColorComboBox, cxCheckBox;

type

  { TdxActivityIndicatorDemoForm }

  TdxActivityIndicatorDemoForm = class(TfmBaseForm)
    ActivityIndicator: TdxActivityIndicator;
    bvlSeparator: TdxBevel;
    ccbArcColor: TcxColorComboBox;
    ccbDotColor: TcxColorComboBox;
    gbArcBased: TcxGroupBox;
    gbCommon: TcxGroupBox;
    gbDotBased: TcxGroupBox;
    gbSettings: TcxGroupBox;
    lbAnimationTime: TcxLabel;
    lbArcColor: TcxLabel;
    lbArcThickness: TcxLabel;
    lbDotColor: TcxLabel;
    lbDotCount: TcxLabel;
    lbDotSize: TcxLabel;
    seAnimationTime: TcxSpinEdit;
    seArcThickness: TcxSpinEdit;
    seDotCount: TcxSpinEdit;
    seDotSize: TcxSpinEdit;
    tcMain: TcxTabControl;

    procedure FormCreate(Sender: TObject);
    procedure seAnimationTimePropertiesChange(Sender: TObject);
    procedure seArcPropertiesChange(Sender: TObject);
    procedure seDotPropertiesChange(Sender: TObject);
    procedure tcMainChange(Sender: TObject);
  private
    FLoading: Boolean;
  end;

var
  dxActivityIndicatorDemoForm: TdxActivityIndicatorDemoForm;

implementation

uses
  dxCoreGraphics;

{$R *.dfm}

{ TdxActivityIndicatorDemoForm }

procedure TdxActivityIndicatorDemoForm.FormCreate(Sender: TObject);
var
  I: Integer;
begin
  tcMain.Properties.Tabs.BeginUpdate;
  try
    tcMain.Properties.Tabs.Clear;
    for I := 0 to GetRegisteredActivityIndicatorProperties.Count - 1 do
      tcMain.Properties.Tabs.Add(GetRegisteredActivityIndicatorProperties.Descriptions[I]);
  finally
    tcMain.Properties.Tabs.EndUpdate;
  end;
  tcMainChange(Sender);
end;

procedure TdxActivityIndicatorDemoForm.seAnimationTimePropertiesChange(Sender: TObject);
begin
  ActivityIndicator.Properties.AnimationTime := seAnimationTime.Value;
end;

procedure TdxActivityIndicatorDemoForm.seArcPropertiesChange(Sender: TObject);
var
  AProperties: TdxActivityIndicatorArcBasedProperties;
begin
  if not FLoading then
    if ActivityIndicator.Properties is TdxActivityIndicatorArcBasedProperties then
    begin
      AProperties := TdxActivityIndicatorArcBasedProperties(ActivityIndicator.Properties);
      AProperties.ArcColor := dxColorToAlphaColor(ccbArcColor.ColorValue);
      AProperties.ArcThickness := seArcThickness.Value;
    end;
end;

procedure TdxActivityIndicatorDemoForm.seDotPropertiesChange(Sender: TObject);
var
  AProperties: TdxActivityIndicatorDotBasedProperties;
begin
  if not FLoading then
    if ActivityIndicator.Properties is TdxActivityIndicatorDotBasedProperties then
    begin
      AProperties := TdxActivityIndicatorDotBasedProperties(ActivityIndicator.Properties);
      AProperties.DotColor := dxColorToAlphaColor(ccbDotColor.ColorValue);
      AProperties.DotCount := seDotCount.Value;
      AProperties.DotSize := seDotSize.Value;
    end;
end;

procedure TdxActivityIndicatorDemoForm.tcMainChange(Sender: TObject);
var
  AArcBasedProperties: TdxActivityIndicatorArcBasedProperties;
  ADotBasedProperties: TdxActivityIndicatorDotBasedProperties;
begin
  if tcMain.TabIndex < 0 then
    Exit;

  FLoading := True;
  try
    ActivityIndicator.PropertiesClassName := GetRegisteredActivityIndicatorProperties.Items[tcMain.TabIndex].ClassName;
    seAnimationTime.Value := ActivityIndicator.Properties.AnimationTime;

    gbArcBased.Visible := ActivityIndicator.Properties is TdxActivityIndicatorArcBasedProperties;
    if gbArcBased.Visible then
    begin
      AArcBasedProperties := TdxActivityIndicatorArcBasedProperties(ActivityIndicator.Properties);
      seArcThickness.Value := AArcBasedProperties.ArcThickness;
      ccbArcColor.ColorValue := dxAlphaColorToColor(AArcBasedProperties.ArcColor);
    end;

    gbDotBased.Visible := ActivityIndicator.Properties is TdxActivityIndicatorDotBasedProperties;
    if gbDotBased.Visible then
    begin
      ADotBasedProperties := TdxActivityIndicatorDotBasedProperties(ActivityIndicator.Properties);
      ccbDotColor.ColorValue := dxAlphaColorToColor(ADotBasedProperties.DotColor);
      seDotCount.Value := ADotBasedProperties.DotCount;
      seDotSize.Value := ADotBasedProperties.DotSize;
    end;
  finally
    FLoading := False;
  end;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressEditors Library/CBuilder/ActivityIndicatorDemo/ActivityIndicatorDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "ActivityIndicatorDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxContainer"
#pragma link "cxControls"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxListView"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "dxActivityIndicator"
#pragma link "cxColorComboBox"
#pragma link "cxDropDownEdit"
#pragma link "cxGroupBox"
#pragma link "cxLabel"
#pragma link "cxMaskEdit"
#pragma link "cxPC"
#pragma link "cxSpinEdit"
#pragma link "cxTextEdit"
#pragma link "dxBevel"
#pragma resource "*.dfm"
TdxActivityIndicatorDemoForm *dxActivityIndicatorDemoForm;
//---------------------------------------------------------------------------
__fastcall TdxActivityIndicatorDemoForm::TdxActivityIndicatorDemoForm(TComponent* Owner)
	: TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TdxActivityIndicatorDemoForm::FormCreate(TObject* Sender)
{
  tcMain->Properties->Tabs->BeginUpdate();
  tcMain->Properties->Tabs->Clear();
  for (int I = 0; I < GetRegisteredActivityIndicatorProperties()->Count; I++)
    tcMain->Properties->Tabs->Add(GetRegisteredActivityIndicatorProperties()->Descriptions[I]);
  tcMain->Properties->Tabs->EndUpdate();
  tcMainChange(Sender);
}
//---------------------------------------------------------------------------
void __fastcall TdxActivityIndicatorDemoForm::seAnimationTimePropertiesChange(TObject* Sender)
{
  ActivityIndicator->Properties->AnimationTime = seAnimationTime->Value;
}
//---------------------------------------------------------------------------
void __fastcall TdxActivityIndicatorDemoForm::seArcPropertiesChange(TObject* Sender)
{
	if (!FLoading)
	{
	  TdxActivityIndicatorArcBasedProperties* AProperties = dynamic_cast<TdxActivityIndicatorArcBasedProperties*>(ActivityIndicator->Properties);
	  AProperties->ArcColor = dxColorToAlphaColor(ccbArcColor->ColorValue);
	  AProperties->ArcThickness = seArcThickness->Value;
	};
}
//---------------------------------------------------------------------------
void __fastcall TdxActivityIndicatorDemoForm::seDotPropertiesChange(TObject* Sender)
{
    if (!FLoading)
	{
	  TdxActivityIndicatorDotBasedProperties* AProperties = dynamic_cast<TdxActivityIndicatorDotBasedProperties*>(ActivityIndicator->Properties);
	  AProperties->DotColor = dxColorToAlphaColor(ccbDotColor->ColorValue);
      AProperties->DotCount = seDotCount->Value;
      AProperties->DotSize = seDotSize->Value;
    };
}
//---------------------------------------------------------------------------
void __fastcall TdxActivityIndicatorDemoForm::tcMainChange(TObject* Sender)
{
  if (tcMain->TabIndex < 0)
    return;

  FLoading = True;
  ActivityIndicator->PropertiesClassName = GetRegisteredActivityIndicatorProperties()->Items[tcMain->TabIndex]->ClassName();
  seAnimationTime->Value = ActivityIndicator->Properties->AnimationTime;

  //# ArcBased
  TdxActivityIndicatorArcBasedProperties* AArcBasedProperties = dynamic_cast<TdxActivityIndicatorArcBasedProperties*>(ActivityIndicator->Properties);
  gbArcBased->Visible = AArcBasedProperties;
  if (gbArcBased->Visible)
  {
	seArcThickness->Value = AArcBasedProperties->ArcThickness;
    ccbArcColor->ColorValue = dxAlphaColorToColor(AArcBasedProperties->ArcColor);
  }

  //# DotBased
  TdxActivityIndicatorDotBasedProperties* ADotBasedProperties = dynamic_cast<TdxActivityIndicatorDotBasedProperties*>(ActivityIndicator->Properties);
  gbDotBased->Visible = ADotBasedProperties;
  if (gbDotBased->Visible)
  {
	ccbDotColor->ColorValue = dxAlphaColorToColor(ADotBasedProperties->DotColor);
    seDotCount->Value = ADotBasedProperties->DotCount;
    seDotSize->Value = ADotBasedProperties->DotSize;
  }
  FLoading = False;
}
//---------------------------------------------------------------------------
